src/Pagerfanta/Adapter/DoctrineORMNativeQueryAdapter.php

```diff
@@ -0,0 +1,122 @@
+ */
+
+namespace Pagerfanta\Adapter;
+
+use Doctrine\ORM\NativeQuery;
+use Doctrine\ORM\Query\ResultSetMapping;
+use Pagerfanta\Exception\InvalidArgumentException;
+
+/**
+ * @author Maxime Veber <nek.dev@gmail.com>
+ */
+class DoctrineORMNativeQueryAdapter implements AdapterInterface
+{
+    /**
+     * @var NativeQuery
+     */
+    private $nativeQuery;
+
+    /**
+     * @var callable
+     */
+    private $countQueryBuilderModifier;
+
+    /**
+     * @var callable
+     */
+    private $sliceQueryBuilderModifier;
+
+    /**
+     * @param NativeQuery  $query                     A DBAL query builder.
+     * @param callable     $countQueryBuilderModifier A callable to modify the query to count.
+     * @param callable     $sliceQueryBuilderModifier A callable to modify the query to slice it (the default should do the job most part of the time).
+     * @throws InvalidArgumentException
+     */
+    public function __construct(NativeQuery $query, $countQueryBuilderModifier = null, $sliceQueryBuilderModifier = null)
+    {
+        if (strpos(strtolower($query->getSQL()), 'select') !== 0) {
+            throw new InvalidArgumentException('Only SELECT queries can be paginated.');
+        }
+
+        if ($countQueryBuilderModifier !== null && !is_callable($countQueryBuilderModifier)) {
+            throw new InvalidArgumentException('The count query builder modifier must be a callable.');
+        }
+        if ($sliceQueryBuilderModifier !== null && !is_callable($sliceQueryBuilderModifier)) {
+            throw new InvalidArgumentException('The slice query builder modifier must be a callable.');
+        }
+
+        $this->nativeQuery = $query;
+        $this->countQueryBuilderModifier = $countQueryBuilderModifier;
+        $this->sliceQueryBuilderModifier = $sliceQueryBuilderModifier;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getNbResults()
+    {
+        $query = $this->cloneQuery();
+        $this->countQueryBuilder($query);
+
+        if ($this->countQueryBuilderModifier !== null) {
+            call_user_func($this->countQueryBuilderModifier, $query);
+        }
+
+        $res = $query->getScalarResult();
+
+        return $res[0]['res'];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSlice($offset, $length)
+    {
+        $query = $this->cloneQuery();
+
+        if ($this->countQueryBuilderModifier !== null) {
+            call_user_func($this->sliceQueryBuilderModifier, $query, $offset, $length);
+        } else {
+            $this->sliceQueryBuilder($query, $offset, $length);
+        }
+
+        return $query->getResult();
+    }
+
+    private function cloneQuery()
+    {
+        $query = clone $this->nativeQuery;
+        $query->setParameters($this->nativeQuery->getParameters());
+
+        return $query;
+    }
+
+    private function countQueryBuilder(NativeQuery $query)
+    {
+        $sql = explode(' FROM ', $query->getSql());
```

Comment:
This will fail when you have a sub-query in the SELECT part.

Reply:
@Nek-, you've explained why you'd like the `countQueryBuilder` method to always be called (even if a modifier is provided), so thank you for that, but I'm not sure you've addressed @stof's concerns that the implentnation of that method will break in a lot of cases.

Reply:
The best answer I can give here is the detection of too complex queries and throw an exception for these cases. Otherwise good thing to do is to build an SQL analyzer.. :')